### PR TITLE
Add Kernel for LagLeadOperator & RollingWindowFeaturizer

### PR DESCRIPTION
**Pull request by wangyems**
Add ORT Kernel for LagLeadOperator
Add ORT Kernel for RollingWindowFeaturizer
keep FeaturizerLibrary commit up to date
fix test cases in ShortGrainDropper kernel
..

### CODE REVIEW

File: onnxruntime/core/graph/featurizers_ops/featurizers_defs.cc [hunk 1]
```diff
@@ -674,6 +676,119 @@ void RegisterLabelEncoderFeaturizerVer1() {
...
+              const auto& target_shape = getInputShape(ctx, 2);
+              if (grains_shape.dim_size() != 2) {
+                fail_shape_inference("Expecting Grains to have 2 dimensions");
+              }
+              if (target_shape.dim_size() != 1) {
+                fail_shape_inference("Expecting Target to have 1 dimensions");
+              }
+              propagateShapeFromInputToOutput(ctx, 1, 0);
+              ONNX_NAMESPACE::TensorShapeProto shape;
+              *shape.add_dim() = grains_shape.dim(0);
```
Comment (yuslepukhin) [inline onnxruntime/core/graph/featurizers_ops/featurizers_defs.cc]: This only requires grains shape to infer but if input 2 does not have shape this does not run.

File: onnxruntime/core/graph/featurizers_ops/featurizers_defs.cc [hunk 2]
```diff
@@ -674,6 +676,119 @@ void RegisterLabelEncoderFeaturizerVer1() {
...
+            if (hasInputShape(ctx, 1) && hasInputShape(ctx, 2)) {
+              const auto& grains_shape = getInputShape(ctx, 1);
+              const auto& target_shape = getInputShape(ctx, 2);
+              if (grains_shape.dim_size() != 2) {
+                fail_shape_inference("Expecting Grains to have 2 dimensions");
+              }
+              if (target_shape.dim_size() != 1) {
+                fail_shape_inference("Expecting Target to have 1 dimensions");
+              }
+              propagateShapeFromInputToOutput(ctx, 1, 0);
```
Comment (yuslepukhin) [inline onnxruntime/core/graph/featurizers_ops/featurizers_defs.cc]: >propagateShapeFromInputToOutput(ctx, 1, 0); [](start = 14, length = 43)

This does not run if input 2 does not have shape

File: onnxruntime/core/graph/featurizers_ops/featurizers_defs.cc [hunk 3]
```diff
@@ -674,6 +676,119 @@ void RegisterLabelEncoderFeaturizerVer1() {
...
+      .TypeAndShapeInferenceFunction(
+          [](ONNX_NAMESPACE::InferenceContext& ctx) {
+            propagateElemTypeFromInputToOutput(ctx, 1, 0);
+            propagateElemTypeFromDtypeToOutput(ctx, ONNX_NAMESPACE::TensorProto_DataType_DOUBLE, 1);
+            if (hasInputShape(ctx, 1) && hasInputShape(ctx, 2)) {
+              const auto& grains_shape = getInputShape(ctx, 1);
+              const auto& target_shape = getInputShape(ctx, 2);
+              if (grains_shape.dim_size() != 2) {
+                fail_shape_inference("Expecting Grains to have 2 dimensions");
+              }
```
Comment (yuslepukhin) [inline onnxruntime/core/graph/featurizers_ops/featurizers_defs.cc]: This check is not performed if input 2 does not have shape.

File: onnxruntime/core/graph/featurizers_ops/featurizers_defs.cc [hunk 4]
```diff
@@ -674,6 +676,119 @@ void RegisterLabelEncoderFeaturizerVer1() {
...
+              if (target_shape.dim_size() != 1) {
+                fail_shape_inference("Expecting Target to have 1 dimensions");
+              }
```
Comment (yuslepukhin) [inline onnxruntime/core/graph/featurizers_ops/featurizers_defs.cc]: This does not check if [R] is the same as input 1

File: onnxruntime/featurizers_ops/cpu/laglead_operator_transformer.cc [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+    std::string* output_grains_data = output_grains_tensor->MutableData<std::string>();
+
+    // Prepare the Output
+    const int64_t output_dim_0 = grains_tensor->Shape()[0];
+
+    double* output_data;
+    bool has_allocate_output_data = false;
+    std::function<void(OutputType)> callback_fn;
+    callback_fn = [&ctx, &output_grains_data, &output_data, &has_allocate_output_data, &output_dim_0](OutputType value) -> void {
+      GrainT output_grains = std::get<GrainT>(value);
```
Comment (yuslepukhin) [inline onnxruntime/featurizers_ops/cpu/laglead_operator_transformer.cc]: >GrainT output_grains = std::get<GrainT>(value); [](start = 6, length = 47)

Does this copy the whole vector?

File: onnxruntime/featurizers_ops/cpu/laglead_operator_transformer.cc [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+      OutputMatrixType output_matrix = std::get<OutputMatrixType>(value);
```
Comment (yuslepukhin) [inline onnxruntime/featurizers_ops/cpu/laglead_operator_transformer.cc]: >OutputMatrixType output_matrix = std::get<OutputMatrixType>(value); [](start = 6, length = 67)

This copies the whole matrix for no reason.

File: onnxruntime/featurizers_ops/cpu/laglead_operator_transformer.cc [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+    TensorShape output_grains_shape = grains_tensor->Shape();
+    Tensor* output_grains_tensor(ctx->Output(0, output_grains_shape));
+    std::string* output_grains_data = output_grains_tensor->MutableData<std::string>();
+
+    // Prepare the Output
+    const int64_t output_dim_0 = grains_tensor->Shape()[0];
+
+    double* output_data;
+    bool has_allocate_output_data = false;
+    std::function<void(OutputType)> callback_fn;
```
Comment (yuslepukhin) [inline onnxruntime/featurizers_ops/cpu/laglead_operator_transformer.cc]: >std::function<void(OutputType)> callback_fn; [](start = 4, length = 44)

This piece of code seems almost totally wrong.
Only floats and doubles are nullable in ORT among the presented datatypes. There are represented as such. Therefore, the output matrix should not require any transformation and should be Mapped/copied directly to the output tensor.

File: onnxruntime/featurizers_ops/cpu/laglead_operator_transformer.cc [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+        output_data = output_tensor->MutableData<double>();
+        has_allocate_output_data = true;
+      }
+      output_grains_data = std::transform(
+                             output_grains.begin(),
+                             output_grains.end(),
+                             output_grains_data,
+                             [](std::string const & s) -> std::string {
+                               return s;
+                             }
```
Comment (yuslepukhin) [inline onnxruntime/featurizers_ops/cpu/laglead_operator_transformer.cc]: What does this do?

Comment (wangyems) [reply onnxruntime/featurizers_ops/cpu/laglead_operator_transformer.cc]: fixed

File: onnxruntime/featurizers_ops/cpu/laglead_operator_transformer.cc (hunk 7 above)
Comment (wangyems) [reply onnxruntime/featurizers_ops/cpu/laglead_operator_transformer.cc]: fixed

**Pull request closed without merging**